tests/FsCodec.SystemTextJson.Tests/UnionConverterTests.fs

```diff
@@ -0,0 +1,361 @@
+//    let s = JsonSerializer.CreateDefault()
+//    let mutable gotError = false
+//    s.Error.Add(fun _ -> gotError <- true)
+
+//    let dJson = """{"case":"CaseD","a":"hi"}"""
+//    use dReader = new StringReader(dJson)
+//    use dJsonReader = new JsonTextReader(dReader)
+//    let d = s.Deserialize<TestDU>(dJsonReader)
+
+//    test <@ (CaseD "hi") = d @>
+//    test <@ false = gotError @>
+
+module ``Unmatched case handling`` =
+    [<Fact>]
+    let ``UnionConverter by default throws on unknown cases`` () =
+        let options = Options.Create(UnionConverter ())
+        let aJson = """{"case":"CaseUnknown"}"""
+        let act () = JsonSerializer.Deserialize<TestDU>(aJson, options)
+
+        fun (e : System.InvalidOperationException) -> <@ -1 <> e.Message.IndexOf "No case defined for 'CaseUnknown', and no catchAllCase nominated" @>
+        |> raisesWith <@ act() @>
+
+    type DuWithCatchAll =
+    | Known
+    | Catchall
+
+    [<Fact>]
+    let ``UnionConverter supports a nominated catchall`` () =
+        let options = Options.Create(UnionConverter<DuWithCatchAll> ("case", "Catchall"))
+        let aJson = """{"case":"CaseUnknown"}"""
+        let a = JsonSerializer.Deserialize<DuWithCatchAll>(aJson, options)
+
+        test <@ Catchall = a @>
+
+    type DuWithMissingCatchAll =
+    | Known
+
+    [<Fact>]
+    let ``UnionConverter explains if nominated catchAll not found`` () =
+        let options = Options.Create(UnionConverter<DuWithMissingCatchAll> ("case", "CatchAllThatCantBeFound"))
+        let aJson = """{"case":"CaseUnknown"}"""
+        let act () = JsonSerializer.Deserialize<DuWithMissingCatchAll>(aJson, options)
+
+        fun (e : System.InvalidOperationException) -> <@ -1 <> e.Message.IndexOf "nominated catchAllCase: 'CatchAllThatCantBeFound' not found" @>
+        |> raisesWith <@ act() @>
+
+    //[<NoComparison>] // Forced by usage of JObject
+    //[<JsonConverter(typeof<UnionConverter>, "case", "Catchall")>]
+    //type DuWithCatchAllWithFields =
+    //| Known
+    //| Catchall of Newtonsoft.Json.Linq.JObject
+
+    //[<Fact>]
+    //let ``UnionConverter can feed unknown values into a JObject for logging or post processing`` () =
+    //    let deserialize json = JsonConvert.DeserializeObject<DuWithCatchAllWithFields>(json, settings)
+    //    let jo =
+    //        trap <@ match deserialize """{"case":"CaseUnknown","a":"s","b":1,"c":true}""" with
+    //                | Catchall jo -> jo
+    //                | x -> failwithf "unexpected %A" x @>
+
+    //    test <@ string jo.["a"]="s"
+    //            && jo.["b"].Type=Newtonsoft.Json.Linq.JTokenType.Integer
+    //            && jo.["c"].Type=Newtonsoft.Json.Linq.JTokenType.Boolean
+    //            && string jo.["case"]="CaseUnknown" @>
+    //    let expected  = "{\r\n  \"case\": \"CaseUnknown\",\r\n  \"a\": \"s\",\r\n  \"b\": 1,\r\n  \"c\": true\r\n}".Replace("\r\n",Environment.NewLine)
+    //    test <@ expected = string jo @>
```

Comment:
TODO: Do we want to have this same test but for JsonElement?

Reply:
Yes, this should be realizable and would definitely be a valuable feature to retain; If it proves tricky and you want to fork it off as an issue that's fine though.
I have not personally thought much about the impl technique, but I imagine Yaron's work in https://github.com/jet/equinox/blob/cosmossdk4/src/Equinox.Cosmos/CosmosJsonSerializer.fs#L10 might provide a useful hint or two

Reply:
It pretty much worked out as-is, except I had to comment out [this bit](https://github.com/jet/FsCodec/pull/43/files#diff-3a3077af061d875164a9a2412e664557R49). I haven't quite though through the implications but all the existing tests (and some more) pass... 😅